networkx/algorithms/bidirectional.py

```diff
@@ -0,0 +1,390 @@
+#    All rights reserved.
+#    BSD license.
+import networkx as nx
+__author__ = """\n""".join(['André Dietrich <dietrich@ivs.cs.uni-magdeburg.de>',
+                            'Sebastian Zug <zug@ivs.cs.uni-magdeburg.de>'])
+__all__ = ['all_simple_paths',
+	   'all_shortest_paths',
+	   'has_path']
+
+
+def all_simple_paths(G, source, target, cutoff=None):
+    """Generate all simple paths in graph G from source to target.
+    
+    A simple path is a path with no repeated nodes.
+    
+    The result paths are returned ordered by there length, starting
+    from the shortest paths.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    source : node
+       Starting node for path
+
+    target : node
+       Ending node for path
+
+    cutoff : integer, optional
+        Depth to stop the search. Only paths of length <= cutoff are returned.
+
+    Returns
+    -------
+    path_generator: generator
+       A generator that produces lists of simple paths, ordered by their
+       length starting from the shortes path. If there are no paths between
+       the source and target within the given cutoff the generator
+       produces no output.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>>
+    >>> G = nx.complete_graph(4)
+    >>> for path in nx.bidirectional.all_simple_paths(G, source=0, target=3):
+    ...     print(path)
+    ...
+    [0, 3]
+    [0, 1, 3]
+    [0, 2, 3]
+    [0, 2, 1, 3]
+    [0, 1, 2, 3]
+    >>> paths = nx.bidirectional.all_simple_paths(G, source=0, target=3, cutoff=2)
+    >>> print(list(paths))
+    [[0, 3], [0, 1, 3], [0, 2, 3]]
+
+    Notes
+    -----
+    This algorithm constructs sequentially two trees, one from the source
+    in direction to the target and one from the target in direction to the
+    source. The leaves of both growing trees are continuously compared, if
+    there are matching leaves in both trees, a new path is identified.
+    
+    This algorithm actually reduces the search space by half, which results
+    in a faster identification of simple paths than in the original
+    "all_simple_paths" algorithm. But it requires more memory, which can
+    have a negative effect on the speed in fully connected graphs.
+    
+              
+                           source    
+                             /\
+                            /\ \
+                           / /\ \
+                          /   /\ \
+                         /\  /\  /\
+                        /  \/   /  \
+                       /    \  /\   \
+                      /_____/_/__\___\_________  meeting points
+                     /\     \    /   /\
+                    /XX\    /  \/   /XX\
+                   /XXXX\   \/  \  /XXXX\
+                  /XXXXXX\  /    \/XXXXXX\
+                 /XXXXXXXX\ \/   /XXXXXXXX\
+                /XXXXXXXXXX\ \  /XXXXXXXXXX\
+               /XXXXXXXXXXXX\ \/XXXXXXXXXXXX\
+              /XXXXXXXXXXXXXX\/XXXXXXXXXXXXXX\
```

Comment:
Neat pic. Let's make sure that Sphinx renders it correctly. We probably will need to mark it as a fixed-width font section.